### ISSUE TITLE
Rethink configurator registration

### ISSUE THREAD

**Issue by amaurym** [regen-network/regen-ledger#301]
**Is your feature request related to a problem? Please describe.**

Modules right now implement this method:

```go
func RegisterServices(configurator servermodule.Configurator, accountKeeper AccountKeeper) {
  configurator.RegisterGenesisHandlers(...)
  configurator.RegisterInvariants(...)
  configurator.RegisterBeginBlock(...) // in the future, maybe?
}
```

It's easy to forget one `Register*`, and it's not clear what all `Register*` possibilities are.

ref: https://github.com/regen-network/regen-ledger/pull/270#discussion_r593151208

**Describe the solution you'd like**

The "old" SDK had all these inside AppModule. It has the advantage of exposing all configurations on a Module, and the skipped ones were just left empty.

One idea could be:
```go
type ModuleConfig interface {
  RegisterMsgServer()
  RegisterQueryServer()
  InitGenesis()
  ExportGenesis()
  BeginBlocker()
}
```

and then `var _ ModuleConfig = impl{}`. For example

```go
func (impl serverImpl) RegisterMsgServer(msgServer gogogrpc.Server) error {
  return group.RegisterMsgServer(msgServer, impl)
}

func (impl serverImpl) InitGenesis(ctx types.Context, cdc codec.JSONMarshaler, data json.RawMessage) []abci.ValidatorUpdate {
  // do logic
}
```

**Describe alternatives you've considered**

- separate interfaces for genesis, invariants... https://github.com/regen-network/regen-ledger/pull/270#discussion_r596653263
- "we could also have one single entry method for the configurator to register genesis handlers, invariants, etc." https://github.com/regen-network/regen-ledger/pull/270#discussion_r596668020

**Additional context**

**Comment by aaronc** [comment]
I think for better compatibility between "old" and "new" (ADR 033) modules, we might want to do something like this:

```go
type Module interface {
  Instantiate(Config) AppModule
}

type Config interface {
  ModuleKey() ModuleKey
  Codec() codec.Marshaler
}

type AppModule interface {
  RegisterMsgServer()
  RegisterQueryServer()
  InitGenesis()
  ExportGenesis()
  BeginBlocker()
}
```

**Comment by robert-zaremba** [comment]
Related: https://github.com/regen-network/regen-ledger/pull/270#discussion_r597948314

The issue is not only about forgetting to register on something - but in general - complexifing the data flow and duplicating functionality (module.Manager vs ModuleManager vs Configurator).

**Comment by robert-zaremba** [comment]
Also, related for refactore: https://github.com/regen-network/regen-ledger/pull/273  and https://github.com/regen-network/regen-ledger/pull/251

**Comment by robert-zaremba** [comment]
> I think for better compatibility between "old" and "new" (ADR 033) modules, we might want to do something like this:

This works. 
@aaronc I assume that in your proposal the `RegisterMsgServer` takes as a parameters `configurator` or `grpc.Server`). In this case we still mix a building responsibility between the module itself (module has to register itself in multiple steps) and 


I was also thinking about a declarative approach:
* a module doesn't register itself, instead it returns required object. With that, the module composition and building process is fully handled by the Manager / Composer. 

    ```
    type AppModule interface {
      MsgSrv() interface{}
      QuerySrv() interface{}
    }
    ```

* unify `types.module.Manager` with `regen/types.module.server.Manager`. Maybe we can create a concept of Composer? This way, maybe, we won't need per module `configurator` object (as it is currently done in `regen/types/module/server/manager.go`).
* remove `regen/types/module.TypeModule`

**Comment by robert-zaremba** [comment]
We discussed this subject during Regen Architecture Meeting -> [notes](https://www.notion.so/regennetwork/Regen-Architecture-Working-Session-e510ed135884474ba20acbe0f2984461).

### Summary

+ We agree that instead of having 4 module managers (BaseManager, module.Manager, regen/Manager, Configurator) we see a need to experiment and create one Manager / Composer
+ The idea would be to: 
  + copy the module.Manager from SDK and rework it without "depending" on other managers. Goal - have only one manager / composer. We can still use Configurator - but not expose it (and not define a public interface).
+ We discussed an abstract factory for a module initializer object.

**Comment by robert-zaremba** [comment]
@aaronc  - after thinking a bit more, I see more need for the module initializer / abstract factory. Main reason for that is test-ability. This would allow to have more options for integration tests.

**Comment by robert-zaremba** [comment]
Maybe it would be easier for tests to mock clients rather than server? If so, client would need to be a constructor parameter, rather than created in constructor as it's done in the [example](https://github.com/regen-network/regen-ledger/blob/aa7aa3e4bfccbac3a0d98a5b332fdb07eb112d6e/x/ecocredit/server/server.go#L37) we were discussing today.

**Comment by robert-zaremba** [comment]
Interesting thing pointed by AiB team is a control of module execution order in BeginBlocker and EndBlocker (control which modules execute first).
cc: @fdymylja , @jgimeno

**Comment by aaronc** [comment]
Yeah I've been thinking about that as well. Rather than an explicit ordering list could a BeginBlocker declare its dependencies and then a dependency graph determines the order deterministically?

**Comment by robert-zaremba** [comment]
Exactly - there is no reason for disallowing that.

**Comment by aaronc** [comment]
Actually in some cases its not explicit dependencies I realize. For instance, the x/upgrade BeginBlocker _always_ needs to run before any other BeginBlocker.
